### PR TITLE
Wm/roberta

### PR DESCRIPTION
**Pull request by willmacd**
Implementation of RoBERTa Transformer model for the HappyTransformer children classes HappyQuestionAnswering, HappyTextClassification and HappyWordPrediction. Merging to V2 this time as opposed to master

Changes:

Implemented RoBERTa transformer model in the QA, TC, and WP classes
Creation of tests for RoBERTa implementation of each class in their respective tests file (i.e. test_qa.py, test_tc.py and test_wp.py)
Creation of a RoBERTa MWP example in the readme_examples.py file

### CODE REVIEW

File: tests/test_wp.py [hunk 1]
```diff
@@ -40,10 +42,20 @@ def test_mwp_basic_albert():
...
     answer = [WordPredictionResult(token_str='.', score=0.8466101884841919)]
-    assert result == answer
\ No newline at end of file
+    assert result == answer
+
+
+def test_mwp_basic_roberta():
+    happy_mwp = HappyWordPrediction("ROBERTA", "roberta-base")
+    result = happy_mwp.predict_mask(
+        "Please pass the salt and <mask>",  # Roberta requires mask to be '<mask>' as opposed to '[MASK]'
```
Comment (ted537) [inline tests/test_wp.py:58]: are we changing the behaviour to require users to insert the correct mask token?

File: happytransformer/happy_word_prediction.py [hunk 2]
```diff
@@ -60,6 +65,10 @@ def predict_mask(self, text, targets=None, top_k=1):
             for answer in result:
                 if answer["token_str"][0] == "▁":
                     answer["token_str"] = answer["token_str"][1:]
+        elif self.model_type == "ROBERTA":
+            for answer in result:
+                if answer["token_str"][0] == "Ġ":
+                    answer["token_str"] = answer["token_str"][1:]
```
Comment (ted537) [inline happytransformer/happy_word_prediction.py:71]: I'm having scalability concerns. We need to decide on where we are going to put this kind of model-specific logic before our base classes get too contrived. 

My intuition wants an `Adapter` for each class that our user-facing classes grab under the hood, that would include the (few) implementation-specific concerns we have

Looks good tho, you followed the pattern we were already using for ALBERT.

I'll probably address this in a separate PR, but I'm interested in thoughts anyone has here.

**Review by ted537 — CHANGES_REQUESTED** [review]

File: tests/test_qa.py [hunk 3]
```diff
@@ -86,9 +89,31 @@ def test_qa_train_effectiveness_bert():
...
+
+    happy_qa = HappyQuestionAnswering("ROBERTA", "roberta-base")
+    before_loss = happy_qa.eval("../data/qa/train-eval.csv").eval_loss
+    happy_qa.train("../data/qa/train-eval.csv")
+    after_loss = happy_qa.eval("../data/qa/train-eval.csv").eval_loss
+    assert after_loss < before_loss
+
+
+def test_qa_test_roberta():
+    happy_qa = HappyQuestionAnswering("ROBERTA", "roberta-base")
```
Comment (EricFillion) [inline tests/test_qa.py:115]: Use a fine-tuned roberta QA model. That's why the results don't make sense. 
 "deepset/roberta-base-squad2"

File: tests/test_wp.py (hunk 1 above)
Comment (EricFillion) [inline tests/test_wp.py:58]: If roberta, replace all [MASK] within the string to <mask>. That way we don't need separate documentation for ROBERTA

**Review by EricFillion — CHANGES_REQUESTED** [review]
Good work, here are some potential changes. I like how you added README examples.

File: tests/test_wp.py (hunk 1 above)
Comment (EricFillion) [reply tests/test_wp.py:58]: I had had the same comment. We are not changing the behaviour.

File: happytransformer/happy_word_prediction.py (hunk 2 above)
Comment (EricFillion) [reply happytransformer/happy_word_prediction.py:71]: I agree we should look into making something a little more scalable. For now a couple if/elif statements is fine.

Comment (ted537) [reply happytransformer/happy_word_prediction.py:71]: When can I do that without treading on toes? Are there any other models we're aiming to add for v2 release?

[the remark this answers is not in the document]
Comment (willmacd) [reply happytransformer/happy_word_prediction.py:71]: So simply create a case for HappyWordPrediction that covers the case that text is empty or is '.' and returns '.' in that case?

[the remark this answers is not in the document]
Comment (EricFillion) [reply happytransformer/happy_word_prediction.py:71]: I made a typo

See what the model outputs when the mask should clearly be a period. If it outputs "</s>", then change it to output "."

[the remark this answers is not in the document]
Comment (EricFillion) [reply happytransformer/happy_word_prediction.py:71]: Never-mind, I did not make a typo, github just changes < /s > to "" for some reason

[the remark this answers is not in the document]
Comment (EricFillion) [reply happytransformer/happy_word_prediction.py:71]: Screen Shot 2021-01-11 at 9 45 48 PM

[the remark this answers is not in the document]
Comment (ted537) [reply happytransformer/happy_word_prediction.py:71]: It's because \</s> is a valid markdown tag. Put your snippet in a codeblock like so to avoid this issue.

\`\`\`
// your code here
\`\`\`

[the remark this answers is not in the document]
Comment (willmacd) [reply happytransformer/happy_word_prediction.py:71]: I ran this: 
image
and received this as the output: 
image
meaning that it doesn't appear to be having issues predicting the character '.'

Please let me know if I am misunderstanding the issue here

Comment (willmacd) [reply happytransformer/happy_word_prediction.py:71]: Is this something that you'd like me to address right now, or is @ted537 taking on this task (in relation to comment above)?

[the remark this answers is not in the document]
Comment (EricFillion) [reply happytransformer/happy_word_prediction.py:71]: Looks fine to me. Maybe they fixed it in a later version or it was a problem with the original Happy Transformer

Comment (EricFillion) [reply happytransformer/happy_word_prediction.py:71]: I missed this comment which is why I'm just replying now. 

I think we're good for models for the first release of v2

File: tests/test_wp.py (hunk 1 above)
Comment (willmacd) [reply tests/test_wp.py]: Updated such that within the 'predict_mask' function the input text replaces instances of '[MASK]' with '<MASK>'; such that the mask token is not changed for the input string yet the text is formatted in a way that RoBERTa likes is

File: examples/question_answering/readme_examples.py [hunk 4]
```diff
@@ -1,10 +1,13 @@
 from happytransformer import HappyQuestionAnswering
+
+
 def example_3_0():
 
     happy_qa_distilbert = HappyQuestionAnswering()  # default
     happy_qa_albert = HappyQuestionAnswering("ALBERT", "mfeb/albert-xxlarge-v2-squad2")
     # good model when using with limited hardware
     happy_qa_bert = HappyQuestionAnswering("BERT", "mrm8488/bert-tiny-5-finetuned-squadv2")
+    happy_qa_roberta = HappyQuestionAnswering("ROBERTA", "deepset/roberta-base-squad2")
```
[the remark this answers is not in the document]
Comment (willmacd) [reply examples/question_answering/readme_examples.py:10]: Added to actual README.md

File: examples/text_classification/readme_examples.py [hunk 5]
```diff
@@ -1,10 +1,11 @@
-
 from happytransformer import HappyTextClassification
 
+
 def example_2_0():
     happy_qa_distilbert = HappyTextClassification()  # default with "distilbert-base-uncased"
     happy_tc_albert = HappyTextClassification(model_type="ALBERT", model_name="albert-base-v2")
     happy_qa_bert = HappyTextClassification("BERT", "bert-base-uncased")
+    happy_qa_roberta = HappyTextClassification("ROBERTA", "roberta-base")
```
[the remark this answers is not in the document]
Comment (willmacd) [reply examples/text_classification/readme_examples.py:8]: Added to actual README.md

[the remark this answers is not in the document]
Comment (willmacd) [reply examples/text_classification/readme_examples.py:8]: and variable names containing 'qa' were changed to 'tc'

[the remark this answers is not in the document]
Comment (ted537) [reply examples/text_classification/readme_examples.py:8]: they weren't tho?

File: examples/word_prediction/readme_examples.py [hunk 6]
```diff
@@ -5,6 +5,7 @@ def example_1_0():
     happy_wp_distilbert = HappyWordPrediction()  # default
     happy_wp_albert = HappyWordPrediction("ALBERT", "albert-base-v2")
     happy_wp_bert = HappyWordPrediction("BERT", "bert-base-uncased")
+    happy_wp_roberta = HappyWordPrediction("ROBERTA", "roberta-base")
```
[the remark this answers is not in the document]
Comment (willmacd) [reply examples/word_prediction/readme_examples.py:8]: Added to actual README.md

File: examples/word_prediction/readme_examples.py [hunk 7]
```diff
@@ -35,8 +36,20 @@ def example_1_3():
     print(result[1].token_str)  # technology
 
 
+def example_1_4():
```
[the remark this answers is not in the document]
Comment (willmacd) [reply examples/word_prediction/readme_examples.py]: Removed

File: tests/test_qa.py (hunk 3 above)
Comment (willmacd) [reply tests/test_qa.py]: Updated to use 'deepset/roberta-base-squad2' as opposed to 'roberta-base'

File: tests/test_wp.py (hunk 1 above)
Comment (willmacd) [reply tests/test_wp.py]: Updated such that within the 'predict_mask' function the input text replaces instances of '[MASK]' with ''; such that the mask token is not changed for the input string yet the text is formatted in a way that RoBERTa likes is

File: examples/text_classification/readme_examples.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (willmacd) [reply examples/text_classification/readme_examples.py]: Sorry, had committed change but hadn't pushed it. Should be fixed now.

**Review by EricFillion — APPROVED** [review]
Good work!

**Review by ted537 — APPROVED** [review]
